pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go

```diff
@@ -1639,6 +1639,581 @@ func TestIsFailure(t *testing.T) {
+			RunObjects:   []v1beta1.RunObject{makeCustomRunStarted(customRuns[0]), makeCustomRunStarted(customRuns[1])},
+		},
+		want: false,
+	}, {
+		name: "one matrixed taskrun running",
+		rpt: ResolvedPipelineTask{
+			PipelineTask: matrixedPipelineTask,
+			TaskRuns:     []*v1beta1.TaskRun{makeStarted(trs[0]), makeSucceeded(trs[1])},
+		},
+		want: false,
+	}, {
+		name: "one matrixed run running",
+		rpt: ResolvedPipelineTask{
+			CustomTask:   true,
+			PipelineTask: matrixedPipelineTask,
+			RunObjects:   []v1beta1.RunObject{makeCustomRunStarted(customRuns[0]), makeCustomRunSucceeded(customRuns[1])},
+		},
+		want: false,
+	}, {
+		name: "matrixed taskruns succeeded",
+		rpt: ResolvedPipelineTask{
+			PipelineTask: matrixedPipelineTask,
+			TaskRuns:     []*v1beta1.TaskRun{makeSucceeded(trs[0]), makeSucceeded(trs[1])},
+		},
+		want: false,
+	}, {
+		name: "one matrixed taskrun succeeded",
+		rpt: ResolvedPipelineTask{
+			PipelineTask: matrixedPipelineTask,
+			TaskRuns:     []*v1beta1.TaskRun{makeSucceeded(trs[0]), makeStarted(trs[1])},
+		},
+		want: false,
+	}, {
+		name: "one matrixed run succeeded",
+		rpt: ResolvedPipelineTask{
+			CustomTask:   true,
+			PipelineTask: matrixedPipelineTask,
+			RunObjects:   []v1beta1.RunObject{makeCustomRunSucceeded(customRuns[0]), makeCustomRunStarted(customRuns[1])},
+		},
+		want: false,
+	}, {
+		name: "matrixed taskruns failed",
+		rpt: ResolvedPipelineTask{
+			PipelineTask: matrixedPipelineTask,
+			TaskRuns:     []*v1beta1.TaskRun{makeFailed(trs[0]), makeFailed(trs[1])},
+		},
+		want: true,
+	}, {
+		name: "matrixed runs failed",
+		rpt: ResolvedPipelineTask{
+			CustomTask:   true,
+			PipelineTask: matrixedPipelineTask,
+			RunObjects:   []v1beta1.RunObject{makeCustomRunFailed(customRuns[0]), makeCustomRunFailed(customRuns[1])},
+		},
+		want: true,
+	}, {
+		name: "one matrixed taskrun failed, one matrixed taskrun running",
+		rpt: ResolvedPipelineTask{
+			PipelineTask: matrixedPipelineTask,
+			TaskRuns:     []*v1beta1.TaskRun{makeFailed(trs[0]), makeStarted(trs[1])},
+		},
+		want: true,
+	}, {
+		name: "one matrixed run failed, one matrixed run running",
+		rpt: ResolvedPipelineTask{
+			CustomTask:   true,
+			PipelineTask: matrixedPipelineTask,
+			RunObjects:   []v1beta1.RunObject{makeCustomRunFailed(customRuns[0]), makeCustomRunStarted(customRuns[1])},
+		},
+		want: true,
+	}, {
+		name: "matrixed taskruns failed: retried",
+		rpt: ResolvedPipelineTask{
+			PipelineTask: withPipelineTaskRetries(*matrixedPipelineTask, 1),
+			TaskRuns:     []*v1beta1.TaskRun{withRetries(makeFailed(trs[0])), withRetries(makeFailed(trs[1]))},
+		},
+		want: true,
+	}, {
+		name: "matrixed runs failed: retries remaining",
+		rpt: ResolvedPipelineTask{
+			CustomTask:   true,
+			PipelineTask: withPipelineTaskRetries(*matrixedPipelineTask, 1),
+			RunObjects:   []v1beta1.RunObject{makeCustomRunFailed(customRuns[0]), makeCustomRunFailed(customRuns[1])},
+		},
+		want: true,
```

Comment:
I'm not sure we want to add tests for custom runs in various states of retrying; we should only test whether the run has completed or not based on its condition succeeded. @XinruZhang worked on refactoring a lot of that logic, maybe she can weigh in here.

Reply:
The retries logic for `TaskRun` is: we only set the Condition as `False` after all retries are exhausted. Same logic applies to `CustomRun`, we ask custom task authors to only set the condition to `False` when all retries are exhausted.

Reply:
It looks like there are a lot of test cases in pipelinerunresolution_test.go and pipelinerunstate_test.go that still test for TaskRun retries status. Xinru, would you mind creating an issue to clean these tests up? It would be a big help!

This PR lgtm other than that.

Reply:
That sounds good, here's the newly created issue: https://github.com/tektoncd/pipeline/issues/6662